File: src/core/EquipmentSlots.js
export const EQUIPMENT_SLOT_KEYS = Object.freeze(['head', 'chest', 'legs', 'feet', 'gloves',
    'shoulders', 'belt', 'neck', 'mainHand', 'offHand', 'ring1', 'ring2', 'trinket1', 'trinket2']);
const slots = new Set(EQUIPMENT_SLOT_KEYS);
const types = new Set(['WEAPON', 'ARMOR', 'ACCESSORY', 'NECK', 'GLOVES']);
const equipmentType = item => item && (!item.type || types.has(item.type));
export const isEquipmentSlot = slot => slots.has(slot);
export const isEquippableItem = item => Boolean(equipmentType(item) &&
    (slots.has(item.slot) || item.slot === 'ring' || item.slot === 'trinket'));
export function itemFitsEquipmentSlot(item, slot) {
    return isEquippableItem(item) && slots.has(slot) && (item.slot === slot ||
        item.slot === 'ring' && ['ring1', 'ring2'].includes(slot) ||
        item.slot === 'trinket' && ['trinket1', 'trinket2'].includes(slot));
}
// Older real-slot gear can omit a slot/type descriptor. Preserve that gear;
// new equip requests still require an explicit matching, supported item slot.
export const isActiveEquipment = (slot, item) => Boolean(slots.has(slot) && equipmentType(item) &&
    (!item.slot || itemFitsEquipmentSlot(item, slot)));

File: server/internal/game/equipment_slots.go
package game

func isEquipmentSlot(slot string) bool {
	switch slot {
	case "head", "chest", "legs", "feet", "gloves", "shoulders", "belt", "neck", "mainHand", "offHand", "ring1", "ring2", "trinket1", "trinket2":
		return true
	}
	return false
}

func equipmentItemType(item Item) bool {
	switch item.Type {
	case "", ItemWeapon, ItemArmor, ItemAccessory, ItemNeck, ItemGloves:
		return true
	}
	return false
}

func itemFitsEquipmentSlot(item Item, slot string) bool {
	return isEquipmentSlot(slot) && equipmentItemType(item) && (item.Slot == slot ||
		item.Slot == "ring" && (slot == "ring1" || slot == "ring2") ||
		item.Slot == "trinket" && (slot == "trinket1" || slot == "trinket2"))
}

func activeEquipmentItem(slot string, item Item) bool {
	// Preserve older gear with omitted descriptors in an actual equipment slot.
	// Unsupported records stay saved for recovery but must not provide power.
	return isEquipmentSlot(slot) && equipmentItemType(item) && (item.Slot == "" || itemFitsEquipmentSlot(item, slot))
}

func activeEquipmentItems(equipment map[string]Item) map[string]Item {
	for slot, item := range equipment {
		if !activeEquipmentItem(slot, item) {
			active := make(map[string]Item, len(equipment))
			for key, value := range equipment {
				if activeEquipmentItem(key, value) {
					active[key] = value
				}
			}
			return active
		}
	}
	return equipment
}
